### ISSUE TITLE
VoiceOver Accessibility

### ISSUE THREAD

**Issue by PeterSaathoff-Harshfield** [Microsoft/vscode#71965]
I brought this up on Twitter last week and got into a bit of a thread with some MS people, but here I am writing something more formal.
VS Code has some issues with the VoiceOver screen reader built into macOS. I think the Electron framework gets in the way here, but VS Code has gotten much better since the last time I tried it sometime last year.
Problems I've run into so far:
There are alerts that appear at the bottom right corner of the window. When VoiceOver focuses on them, they are reported as "notification toast." VoiceOver doesn't read anything more than "notification toast" for each one. They can't be interacted with, and therefore can't be read or dismissed with VoiceOver. Also, "toast" is maybe not a meaningful word for many people in this context even though programmers may know what a toast is.
I access the settings screen with the standard "cmd + ,". It has a search field at the top. There is placeholder text in this field, "Search settings" but this is a separate text element. I would expect placeholder text to somehow be part of the field rather than a separate element.
On the left side of the settings screen is a section called "commonly used." This section can't be read or interacted with by VoiceOver. It is reported as, "empty table."
After this comes, "settings list focus helper, group," which seems to outline most of the window and contains no other elements.
Next is "settings, table." This appears to contain the bulk of the settings. Interacting with this area does nothing. VoiceOver knows about no elements within. Interestingly, using tab at this point will cause VoiceOver to move focus to controls within this area, but only editable controls and links. None of the static text is accessible.
Below the main settings area is another "settings list focus helper" that appears to do nothing.
At the bottom of the window is a bar that VoiceOver reports as, "content info." The elements in this bar are navigated to by VoiceOver in the opposite order than they appear. When I activate notifications, I hear something spoken about the Arduino package being recommended, but then VoiceOver focus lands inside a "notifications list" that I can't interact with.
Now I'm out of settings editing a text file. There is a tab for the document, and this tab includes a "tab actions" group with only a "close" button inside it.
When I interact with the main text field, most of the standard text editing operations work. I can move among lines with up/down arrow keys, but attempting to move among them using "control + option + up/down" doesn't work correctly. This should do the same thing as just up/down arrows, but instead seems to cause VoiceOver to want to move focus to elements below/above because it plays the "bonk" sound to indicate that's not possible. In any other text editor or text field, the VoiceOver Focus box visually surrounds the current line of text. In this one, the VoiceOver focus box seems to become a thin line floating up above the text field. The fact that "control + option + up/down" doesn't work and the focus box is in the wrong place must be connected.
I'm sure I haven't touched most of the advanced features of VS Code, but I've run into all this while just doing basic text editing.

**Comment by isidorn** [comment]
@PeterSaathoff-Harshfield thanks a lot for your feedback.
Since we prefer smaller issues compared to one giant one I went ahead and split this one in multiple smaller ones. Can you please start following those and answer potential questions we might have.
Also more feedback is very welcome! Just small reproducable issues are the best.

Notifications: https://github.com/Microsoft/vscode/issues/72077
Commonly used in settings: https://github.com/Microsoft/vscode/issues/72078
Settings editor structure: https://github.com/Microsoft/vscode/issues/72079

Regarding your feedback for ctrl + fn + up/down not working. I can not reproduce this. When I press this button and have VoiceOver enabled, VoiceOver does not read anything, since the cursor does not move, only the view port changes since it scrolls. I think this is expacted behavior. Can you please provide more details regarding this here, so I know if I should open a sepearte issue.

**Comment by tayarndt** [comment]
Another issue I am not sure you are having is if you are in a file with many blank lines in between different sections voiceover. reads the text as many blank lines were between it. For example, If I have a file with 2 blank lines than vs code would read the text twice. please look at the issue  https://github.com/Microsoft/vscode/issues/72570#issue-434852963
Let me know if you all have had this issue? I think it is weird and could be electron related?

**Comment by tayarndt** [comment]
That could be the wrong issue number but not sure.

**Comment by PeterSaathoff-Harshfield** [comment]
Isidorn, I think you had the wrong keys for moving up/down a line in the text field. Control + option are the standard VoiceOver modifier keys. Holding these down while pressing up/down arrows in a text field should just move up or down by a line. If you try this in VS Code, you will find it's inconsistent.
In the first place, I can almost never "interact with" the text field, so I don't think VoiceOver even knows it's in the text field. So, when I use control + option + up/down, VoiceOver makes a "ding" sound to indicate I can't move.
Perhaps every 5 restarts of VS Code and VoiceOver, I actually can get VoiceOver to "interact with" the text field, and control + option + up/down do work.
I've tried to record a video of this behavior several times, but it acts differently each time I make a recording.

**Comment by tayarndt** [comment]
That is so weird! I do the voiceover modifier keys and get the behavior you describe.

Taylor Arndt
Accessibility specialist
University computing center 2nd floor
Summer hours
Monday-Friday 9:00-5:00
work email oit-stl@wmich.edu

**Comment by tayarndt** [comment]
Have you found a workaround? How do you move by lines? I have tried also using the arrow keys with little more success.

**Comment by isidorn** [comment]
@PeterSaathoff-Harshfield we are using different keyboard shortcuts for this.
This should work if you press `ctrl + pageUp` and `ctrl + pageDown`.
Or if your mac keyboard does not have pageUp pageDown you can press `ctrl + fn + up` and `ctrl + fn + down`
This can be configured via Preferences > Keyboard Shortcuts > `scrollLineDown` and `scrollLineUp`
Configure any keybindings you would want there.

Let me know if something does not work and I can reopen this issue.
Closing for now since I think we have tackled all your points. Thanks

**Comment by PeterSaathoff-Harshfield** [comment]
Control + page up/page down just scrolls the text field by one line.
What I'm referring to is moving the insertion point AND VoiceOver cursor up and down among lines. This does not work on version 1.33.1.
